apps/web-giddh/src/app/accounting/journal-voucher/voucher/receipt-entry-modal/receipt-entry-modal.component.ts

```diff
@@ -0,0 +1,515 @@
+     *
+     * @memberof ReceiptEntryModalComponent
+     */
+    public addNewEntry(): void {
+        if (this.totalEntries === 0 || (this.receiptEntries[this.totalEntries - 1] && this.receiptEntries[this.totalEntries - 1] !== undefined && parseFloat(this.receiptEntries[this.totalEntries - 1].amount) > 0)) {
+            let getAdjustmentTypes = this.prepareAdjustmentTypes("add");
+
+            this.receiptEntries[this.totalEntries] = {
+                allowedTypes: getAdjustmentTypes,
+                type: (this.advanceReceiptExists) ? 'advanceReceipt' : 'receipt',
+                //note: '',
+                tax: {
+                    name: '',
+                    uniqueName: '',
+                    percent: 0,
+                    value: 0
+                },
+                invoice: {
+                    number: '',
+                    date: '',
+                    amount: 0,
+                    uniqueName: '',
+                    type: ''
+                },
+                amount: 0
+            }
+            this.totalEntries++;
+        }
+    }
+
+    /**
+     * This will get called on enter/tab in adjustment amount field
+     *
+     * @param {KeyboardEvent} event
+     * @param {*} entry
+     * @memberof ReceiptEntryModalComponent
+     */
+    public validateAmount(event: KeyboardEvent, entry: any): void {
+        if ((event.keyCode === 9 || event.keyCode === 13) && this.transaction && this.transaction.amount) {
+            this.validateEntry(entry);
+        }
+    }
+
+    /**
+     * This will validate the adjustment entry
+     *
+     * @param {*} entry
+     * @returns {*}
+     * @memberof ReceiptEntryModalComponent
+     */
+    public validateEntry(entry: any): any {
+        if (!entry.amount) {
+            this.toaster.clearAllToaster();
+            this.toaster.errorToast(this.invalidAmountErrorMessage);
+            this.isValidForm = false;
+            return;
+        } else if (isNaN(parseFloat(entry.amount)) || entry.amount <= 0) {
+            this.toaster.clearAllToaster();
+            this.toaster.errorToast(this.invalidAmountErrorMessage);
+            this.isValidForm = false;
+            return;
+        }
+
+        if (entry.type === "againstReference" && !entry.invoice.uniqueName) {
+            this.toaster.clearAllToaster();
+            this.toaster.errorToast(this.invoiceErrorMessage);
+            this.isValidForm = false;
+            return;
+        }
+
+        if (entry.amount && !isNaN(parseFloat(entry.amount)) && entry.tax.percent) {
+            entry.tax.value = parseFloat(entry.amount) / entry.tax.percent;
+        } else {
+            entry.tax.value = 0;
+        }
+
+        let receiptTotal = 0;
+
+        this.receiptEntries.forEach(receipt => {
+            receiptTotal += parseFloat(receipt.amount);
+        });
+
+        if (receiptTotal < this.transaction.amount) {
+            if (entry.type === "againstReference") {
```

Comment:
Consider using enum for entry type with values `againstReference`, `advanceReceipt`, `receipt`